[Documentation] Yolov5 (#830)
neuralmagic/sparseml: src/sparseml/yolov5/README.md

# SparseML YOLOv5 Integration
This directory combines the SparseML recipe-driven approach with the 
[ultralytics/yolov5](https://github.com/ultralytics/yolov5) repository.

> By integrating the robust training flows in the `yolov5` repository with the SparseML code base, we enable model sparsification techniques on the popular [YOLOv5 architecture](https://github.com/ultralytics/yolov5/issues/280)

Author: I think the architecture is too tangential to the content here to include directly. In this case the reference was created as an issue, but not meant to be an "issue" rather than a thread of documentation and discussion. Definitely a weird quirk in the Ultralytics repo, but I think it's a choice we have to live with. Also for context, this part of the content is borrowed from the previous YoloV5 documentation written by us.

> By integrating the robust training flows in the `yolov5` repository with the SparseML code base, we enable model sparsification techniques on the popular [YOLOv5 architecture](https://github.com/ultralytics/yolov5/issues/280)

Reviewer: Ok, I agree with the logic.

as well as the updated [YOLOV5x6 architecture](https://github.com/ultralytics/yolov5/releases/tag/v5.0),
creating smaller and faster deployable versions.
The techniques include, but are not limited to:
- Pruning
- Quantization
- Pruning + Quantization
- Sparse Transfer Learning

After training, the model can be deployed with Neural Magic's DeepSparse Engine. The engine enables inference with GPU-class performance directly on your CPU.

This integration enables spinning up one of the following end-to-end functionalities:
- **Sparsification of YOLOv5 Models** - easily sparsify any of the YOLOV5 and YOLOV5x6 models, from YOLOv5n to YOLOv5x models. 
- **Sparse Transfer Learning** - fine-tune a sparse backbone model (or use one of our [sparse pre-trained models](https://sparsezoo.neuralmagic.com/?domain=cv&sub_domain=detection&page=1)) on your own, private dataset.

## Installation

```pip install sparseml[torchvision]```

> Note: YOLOV5 will not immediately install with this command. Instead, a sparsification-compatible version of YOLOv5 will install on the first invocation of the YOLOv5 code in SparseML.

Reviewer: ```suggestion
Note: YOLOv5 will not immediately install with this command. Instead, a sparsification-compatible version of YOLOv5 will install on the first invocation of the YOLOv5 code in SparseML.
```

nit

## Tutorials

- [Sparsifying YOLOv5 Using Recipes](https://github.com/neuralmagic/sparseml/blob/main/integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md)
- [Sparse Transfer Learning With YOLOv5](https://github.com/neuralmagic/sparseml/blob/main/integrations/ultralytics-yolov5/tutorials/yolov5_sparse_transfer_learning.md)

## Quick Tour

### Recipes

> Recipes encode the instructions and hyperparameters for sparsifying a model using modifiers to the training process.

Reviewer: When writing my docs I implicitly assumed that the user knows what recipes are. My reasoning was, that it is very probable that the user probably would get exposed to recipes (e.g. while reading the docs), before going into the codebase. But I guess, mentioning the recipes shortly is quite valuable and elegant.

The modifiers can range from pruning and quantization to learning rate and weight decay.
When appropriately combined, it becomes possible to create highly sparse and accurate models.

This integration adds a `--recipe` argument to the [`train.py` script](https://github.com/neuralmagic/yolov5/blob/master/train.py).
The argument loads an appropriate recipe while preserving the rest of the training pipeline.

> Popular recipes used with this argument are found in the [`recipes` folder](./recipes).

Reviewer 2: ```suggestion
Popular recipes used with this argument can be found in the [`recipes` folder](./recipes).
```

> Popular recipes used with this argument are found in the [`recipes` folder](./recipes).

Reviewer 2: Noting: Could you confirm if relative links work with Sphinx? I remember @jeanniefinks noticed they weren't

Otherwise, all other arguments and functionality remain the same as the original repository.

### SparseZoo

Neural Magic’s ML team creates sparsified models that allow anyone to plug in their data and leverage pre-sparsified models from the SparseZoo.
Select a yolov5 model from the [SparseZoo](https://sparsezoo.neuralmagic.com/?repo=ultralytics&page=1).

### SparseML CLI

The SparseML installation provides a CLI for running YOLOv5 scripts with SparseML capability. The full set of commands is included below

```bash
sparseml.yolov5.train
sparseml.yolov5.val
sparseml.yolov5.export_onnx
sparseml.yolov5.val_onnx
```

Appending the `--help` argument displays a full list of options for the command:
```bash
sparseml.yolov5.train --help

> ```

Author: I considered adding it, but --help ends up being too long because there are quite a lot of arguments. I think in this case it's best to leave it out

## Installation

```pip install sparseml[torchvision]```

Note: YOLOv5 will not immediately install with this command. Instead, a sparsification-compatible version of YOLOv5 will install on the first invocation of the YOLOv5 code in SparseML.

## Tutorials

- [Sparsifying YOLOv5 Using Recipes](https://github.com/neuralmagic/sparseml/blob/main/integrations/ultralytics-yolov5/tutorials/sparsifying_yolov5_using_recipes.md)
- [Sparse Transfer Learning With YOLOv5](https://github.com/neuralmagic/sparseml/blob/main/integrations/ultralytics-yolov5/tutorials/yolov5_sparse_transfer_learning.md)

## Quick Tour

### Recipes

Recipes encode the instructions and hyperparameters for sparsifying a model using modifiers to the training process.
The modifiers can range from pruning and quantization to learning rate and weight decay.
When appropriately combined, it becomes possible to create highly sparse and accurate models.

### SparseZoo

Neural Magic’s ML team creates sparsified models that allow anyone to plug in their data and leverage pre-sparsified models from the SparseZoo.

> Select a yolov5 model from the [SparseZoo](https://sparsezoo.neuralmagic.com/?repo=ultralytics&page=1).

Reviewer 3: ```suggestion
Select a YOLOv5 model from the [SparseZoo](https://sparsezoo.neuralmagic.com/?repo=ultralytics&page=1).
```

  --hyp data/hyps/hyp.finetune.yaml \
  --recipe zoo:cv/detection/yolov5-s/pytorch/ultralytics/coco/pruned-aggressive_96
```

## Once the Training is Done...

### Exporting the Sparse Model to ONNX
The DeepSparse Engine accepts ONNX formats and is engineered to significantly speed up inference on CPUs for the sparsified models from this integration.

The SparseML installation provides a `sparseml.yolov5.export_onnx` command that you can use to load the training model folder and create a new `model.onnx` file within. The export process is modified such that the quantized and pruned models are corrected and folded properly. Be sure the `--weights` argument points to your trained model. 
```bash
sparseml.yolov5.export_onnx \
    --weights path/to/weights.pt \
    --dynamic 
```

### DeepSparse Engine Benchmarking and Deployment

Once the model is exported to the ONNX format, it is ready for deployment with the DeepSparse Engine. 

To benchmark the model, you can run

```bash
deepsparse.benchmark path/to/model.onnx
```

To run validation in the DeepSparse Engine, you can run

```bash
sparseml.yolov5.val_onnx path/to/model.onnx --data coco128.yaml
```

> To learn more about the engine and deploying your model, refer to the [appropriate documentation in the DeepSparse repository](https://github.com/neuralmagic/deepsparse/tree/main/examples/ultralytics-yolo)

Reviewer 4: let's point to the new deepsparse doc instead
